### PR TITLE
[e2e] Check labels documentation

### LINKED ISSUE #1078: e2e: check labels documentation
As @lilic [pointed out](https://github.com/kubernetes/kube-state-metrics/pull/996#pullrequestreview-351396437) recently, CI checks thats every metric is documented (`make doccheck`), but there's no check on labels documentation.

That led to PR where labels documentation was not updated and CI did not catch this (#996).

Actually, it's more difficult to check labels:

- metric names are extracted from source code by a static analysis (with regex `"(kube_[^"]+)"`); labels are not easily found by basic static analysis; 
- some labels have a dynamic name (e.g. `label_JOB_LABEL`)

What do you think of using e2e to check labels documentation? We already check metrics from `/metrics` endpoint; we could also list labels for each metric and check if they're in documentation (even if it is not exhaustive).

### PR DESCRIPTION
**Pull request by olivierlemasle**
**What this PR does / why we need it**:
This PR adds a new e2e test, checking that all metrics and labels found in `/metrics` endpoint are correctly documented (currently, metrics documentation is already checked with `make doccheck`, but labels are not validated).

When part of the label name is documented in uppercase (e.g. `label_DAEMONSET_LABEL`), we will consider that by convention, it means that the uppercase part is dynamic (checked as a wildcard).

NB: we also remove test manifest for `Lease`; as explained in the commit message, this was not necessary and caused an (ignored) error at the end of e2e tests.

**Which issue(s) this PR fixes**:
Fixes #1078

### CODE REVIEW

**Comment by olivierlemasle** [comment]
CI failures are actually good news:

<details>

```
=== RUN   TestDocumentation
---- FAIL: TestDocumentation (0.08s)
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_spec_min_replicas are: hpa, namespace
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_spec_max_replicas are: hpa, namespace
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_status_current_replicas are: hpa, namespace
-    main_test.go:112: Metric kube_validatingwebhookconfiguration_created not found in documentation.
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_allocatable are: node, resource
-    main_test.go:112: Metric kube_pod_start_time not found in documentation.
-    main_test.go:112: Metric kube_mutatingwebhookconfiguration_created not found in documentation.
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_status_condition are: hpa, namespace, condition, status
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_status_condition are: hpa, namespace, condition, status
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_status_condition are: hpa, namespace, condition, status
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label unit not found in documentation. Documented labels for metric kube_node_status_capacity are: node, resource
-    main_test.go:131: Label label_app_kubernetes_io_name not found in documentation. Documented labels for metric kube_replicaset_labels are: replicaset, namespace
-    main_test.go:131: Label label_app_kubernetes_io_version not found in documentation. Documented labels for metric kube_replicaset_labels are: replicaset, namespace
-    main_test.go:131: Label label_pod_template_hash not found in documentation. Documented labels for metric kube_replicaset_labels are: replicaset, namespace
-    main_test.go:131: Label label_k8s_app not found in documentation. Documented labels for metric kube_replicaset_labels are: replicaset, namespace
-    main_test.go:131: Label label_pod_template_hash not found in documentation. Documented labels for metric kube_replicaset_labels are: replicaset, namespace
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_metadata_generation are: hpa, namespace
-    main_test.go:112: Metric kube_mutatingwebhookconfiguration_info not found in documentation.
-    main_test.go:112: Metric kube_validatingwebhookconfiguration_info not found in documentation.
-    main_test.go:131: Label label_k8s_app not found in documentation. Documented labels for metric kube_deployment_labels are: deployment, namespace
-    main_test.go:131: Label label_app_kubernetes_io_name not found in documentation. Documented labels for metric kube_deployment_labels are: deployment, namespace
-    main_test.go:131: Label label_app_kubernetes_io_version not found in documentation. Documented labels for metric kube_deployment_labels are: deployment, namespace
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label phase not found in documentation. Documented labels for metric kube_namespace_status_phase are: namespace, status
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_spec_target_metric are: hpa, namespace, metric_name, metric_target_type
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_status_desired_replicas are: hpa, namespace
-    main_test.go:112: Metric kube_validatingwebhookconfiguration_metadata_resource_version not found in documentation.
-    main_test.go:112: Metric kube_mutatingwebhookconfiguration_metadata_resource_version not found in documentation.
-    main_test.go:112: Metric kube_node_spec_unschedulable not found in documentation.
-    main_test.go:131: Label horizontalpodautoscaler not found in documentation. Documented labels for metric kube_horizontalpodautoscaler_labels are: hpa, namespace
-FAIL
-FAIL	k8s.io/kube-state-metrics/tests/e2e	0.121s
-FAIL
```
</details>

It shows the following errors, that I will fix in a new commit:

- Metrics `kube_pod_start_time` and `kube_node_spec_unschedulable` do not have a stability stage (STABLE, EXPERIMENTAL, etc.)
- For metrics `kube_node_status_capacity` and `kube_node_status_allocatable`, in the definition of label `unit`, characters `` ` `` and `=` have been exchanged.
- Documentation of metric `kube_namespace_status_phase` misses label `phase`
- Metrics `kube_deployment_labels` and `kube_replicaset_labels` miss dynamic label `label_*`
- Metrics `kube_horizontalpodautoscaler_*` have a label `horizontalpodautoscaler`, whereas the old label `hpa` is still documented.
- Documentation files for `mutatingwebhookconfiguration` and `validatingwebhookconfiguration` do not end by `-metrics.md` like other metric documentation files.

/hold

**Comment by olivierlemasle** [comment]
Added commit with documentation fixes; it resolves e2e test failures.
/unhold

File: tests/manifests/lease.yaml [hunk 1]
```diff
@@ -1,14 +0,0 @@
-apiVersion: coordination.k8s.io/v1
```
Comment (tariq1890) [inline tests/manifests/lease.yaml]: Why delete this?

Comment (tariq1890) [reply tests/manifests/lease.yaml]: Nvm, I saw the commit message. 

Sounds good to me :)

File: tests/e2e/main_test.go [hunk 2]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+// It returns a map where keys are metric names, and values are slices of label names,
+// and an error in case of failure.
+// By convention, UPPER_CASE parts in label names denotes wilcard patterns, used for dynamic labels.
+func getLabelsDocumentation() (map[string][]string, error) {
+	documentedMetrics := map[string][]string{}
+
+	docPath := "../../docs/"
+	docFiles, err := ioutil.ReadDir(docPath)
+	if err != nil {
+		return documentedMetrics, fmt.Errorf("failed to read documentation directory: %w", err)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]:
```suggestion
		return documentedMetrics, errors.Wrap(err, "failed to read documentation directory")
```

File: tests/e2e/main_test.go [hunk 3]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+
+	for _, file := range docFiles {
+		if file.IsDir() || !fileRe.MatchString(file.Name()) {
+			continue
+		}
+
+		filePath := path.Join(docPath, file.Name())
+		f, err := os.Open(filePath)
+		if err != nil {
+			return documentedMetrics, fmt.Errorf("cannot read file %v: %w", filePath, err)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: `%s` works for string slice too.
```suggestion
			return documentedMetrics, errors.Wrap("cannot read file %s", filePath)
```

File: tests/e2e/main_test.go [hunk 4]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+				continue
+			}
+			metric := params[1]
+			labelsDoc := params[2]
+
+			labels := labelsRe.FindAllStringSubmatch(labelsDoc, -1)
+			labelPatterns := make([]string, len(labels))
+			for i, l := range labels {
+				if len(l) <= 1 {
+					return documentedMetrics, fmt.Errorf("Label documentation %v did not match regex", labelsDoc)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]:
```suggestion
					return documentedMetrics, errors.Errorf("Label documentation %s did not match regex", labelsDoc)
```

**Comment by tariq1890** [comment]
This is so awesome. Thanks @olivierlemasle :)!

File: tests/e2e/main_test.go [hunk 5]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+}
+
+// getLabelsDocumentation is a helper function that gets metric mabels documentation.
+// It returns a map where keys are metric names, and values are slices of label names,
+// and an error in case of failure.
+// By convention, UPPER_CASE parts in label names denotes wilcard patterns, used for dynamic labels.
+func getLabelsDocumentation() (map[string][]string, error) {
+	documentedMetrics := map[string][]string{}
+
+	docPath := "../../docs/"
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: Are there any go libraries from `stdlib` to get the directory the project root?

Comment (olivierlemasle) [reply tests/e2e/main_test.go]: Hmm... I suppose we could use https://stackoverflow.com/a/38644571/625158, that uses [runtime.Caller](https://golang.org/pkg/runtime/#Caller), but is it better?

File: tests/e2e/main_test.go (hunk 2 above)
Comment (olivierlemasle) [reply tests/e2e/main_test.go]: Done

File: tests/e2e/main_test.go (hunk 3 above)
Comment (olivierlemasle) [reply tests/e2e/main_test.go]: Ok, thanks

File: tests/e2e/main_test.go (hunk 5 above)
Comment (tariq1890) [reply tests/e2e/main_test.go]: I think we can go with that :). From the looks of it, I don't see any problems.

File: tests/e2e/main_test.go [hunk 6]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
 	}
 }
+
+func TestDocumentation(t *testing.T) {
+	labelsDocumentation, err := getLabelsDocumentation()
+	if err != nil {
+		t.Fatal("Cannot get labels documentation", err)
+	}
+
+	buf := &bytes.Buffer{}
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: Not a strong opinion. But this is just my preference in the interest of semantics.

```suggestion
	var buf *bytes.Buffer
```

File: tests/e2e/main_test.go [hunk 7]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+// It returns a map where keys are metric names, and values are slices of label names,
+// and an error in case of failure.
+// By convention, UPPER_CASE parts in label names denotes wilcard patterns, used for dynamic labels.
+func getLabelsDocumentation() (map[string][]string, error) {
+	documentedMetrics := map[string][]string{}
+
+	docPath := "../../docs/"
+	docFiles, err := ioutil.ReadDir(docPath)
+	if err != nil {
+		return documentedMetrics, errors.Wrap(err, "failed to read documentation directory")
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: I think we should return nil in the cases of error. That is the accepted pattern I believe
```suggestion
		return nil, errors.Wrap(err, "failed to read documentation directory")
```

File: tests/e2e/main_test.go [hunk 8]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+
+	for _, file := range docFiles {
+		if file.IsDir() || !fileRe.MatchString(file.Name()) {
+			continue
+		}
+
+		filePath := path.Join(docPath, file.Name())
+		f, err := os.Open(filePath)
+		if err != nil {
+			return documentedMetrics, errors.Wrapf(err, "cannot read file %s", filePath)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: Let's return nil please

File: tests/e2e/main_test.go [hunk 9]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+				continue
+			}
+			metric := params[1]
+			labelsDoc := params[2]
+
+			labels := labelsRe.FindAllStringSubmatch(labelsDoc, -1)
+			labelPatterns := make([]string, len(labels))
+			for i, l := range labels {
+				if len(l) <= 1 {
+					return documentedMetrics, errors.Errorf("Label documentation %s did not match regex", labelsDoc)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]: return nil here as well

File: tests/e2e/main_test.go [hunk 10]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+						t.Errorf("Cannot compile pattern %v: %v", labelPattern, err)
+						continue
+					}
+					if re.MatchString(labelName) {
+						labelNameMatched = true
+						break
+					}
+				}
+				if !labelNameMatched {
+					t.Errorf("Label %v not found in documentation. Documented labels for metric %v are: %v",
```
Comment (tariq1890) [inline tests/e2e/main_test.go]:
```suggestion
					t.Errorf("Label %s not found in documentation. Documented labels for metric %s are: %s",
```

File: tests/e2e/main_test.go [hunk 11]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+			continue
+		}
+		for _, m := range metricFamily.Metric {
+			for _, l := range m.Label {
+				labelName := l.GetName()
+				labelNameMatched := false
+				for _, labelPattern := range acceptedLabelNames {
+					re, err := regexp.Compile(labelPattern)
+					if err != nil {
+						t.Errorf("Cannot compile pattern %v: %v", labelPattern, err)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]:
```suggestion
						t.Errorf("Cannot compile pattern %v: %w", labelPattern, err)
```

File: tests/e2e/main_test.go [hunk 12]
```diff
@@ -77,3 +84,110 @@ func TestLintMetrics(t *testing.T) {
...
+	if err != nil {
+		t.Fatal("Cannot decode metrics", err)
+	}
+
+	for _, metricFamily := range metricFamilies {
+		metric := metricFamily.GetName()
+
+		acceptedLabelNames, ok := labelsDocumentation[metric]
+		if !ok {
+			t.Errorf("Metric %v not found in documentation.", metric)
```
Comment (tariq1890) [inline tests/e2e/main_test.go]:
```suggestion
			t.Errorf("Metric %s not found in documentation.", metric)
```

File: tests/e2e/main_test.go (hunk 5 above)
Comment (olivierlemasle) [reply tests/e2e/main_test.go]: I've just tested adding:
```go
_, b, _, _ = runtime.Caller(0)
```
Using `make e2e` (from the root path of the project), `b` is actually the absolute path of `./tests/e2e/main_test.go`. So I cannot get the documentation path with this method, without using `../../docs/`.

File: tests/e2e/main_test.go (hunk 7 above)
Comment (olivierlemasle) [reply tests/e2e/main_test.go]: Sure!

**Comment by olivierlemasle** [comment]
@tariq1890 PTAL thanks!

**Comment by tariq1890** [comment]
/lgtm 

Thanks again @olivierlemasle :)

**Pull request merged**